contracts/oracles/CurveLPOracle.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: bsl-1.1
+
+/*
+  Copyright 2020 Unit Protocol: Artem Zakharov (az@unit.xyz).
+*/
+pragma solidity ^0.7.1;
+
+import "./OracleSimple.sol";
+import "../helpers/ERC20Like.sol";
+import "./OracleRegistry.sol";
+
+interface CurveProvider {
+    function get_registry() external view returns (address);
+}
+
+interface CurveRegistry {
+    function get_pool_from_lp_token(address) external view returns (address);
+    function get_virtual_price_from_lp_token(address) external view returns (uint);
+}
+
+/**
+ * @title CurveLPOracle
+ * @dev Oracle to quote curve LP tokens
+ **/
+contract CurveLPOracle is OracleSimple {
+
+    uint public constant Q112 = 2 ** 112;
+    uint public constant PRECISION = 1e18;
+
+    // CurveProvider contract
+    CurveProvider public immutable curveProvider;
+
+    /**
+     * @param _curveProvider The address of the Curve Provider. Mainnet: 0x0000000022D53366457F9d5E68Ec105046FC4383
+     **/
+    constructor(address _curveProvider) {
+        require(_curveProvider != address(0), "Unit Protocol: ZERO_ADDRESS");
+        curveProvider = CurveProvider(_curveProvider);
+    }
+
+    // returns Q112-encoded value
+    function assetToUsd(address asset, uint amount) public override view returns (uint) {
+        if (amount == 0) return 0;
+        CurveRegistry cR = CurveRegistry(curveProvider.get_registry());
+        require(cR.get_pool_from_lp_token(asset) != address(0), "Unit Protocol: NOT_A_CURVE_LP");
+
+        uint price_q112 = cR.get_virtual_price_from_lp_token(asset) * Q112 / PRECISION;
```

Comment:
- [ ] `get_virtual_price_from_lp_token` is not a price at all (despite the name, lol), it's a ratio. See for example the price for the renBTC/wBTC/sBTC LP 0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3. Therefore, it can't be used that way, even for usd-like assets.
- [ ] it seems to me that the pool doesn't have any idea about it's dollar value, so neither `get_virtual_price` nor any other function would work for the task. The only option I see is to assess the pool balance in each of the pool `coins`.
- [ ] the naive one-shot solution would still be prone to a flash loan manipulation in the case where one of the coins deviated from the peg.

Reply:
get_virtual_price has a meaning of price of the pool _if prices of all assets involved change to 1.0 smoothly_.

If you worry about one of assets being irreversibly rekt (like usdt), you need to take `price = virtual_price * min(*token_prices) // 10**18`, where `*token_prices` for individual tokens are taken from chainlink or whatnot

Reply:
On the other hand, if the worry is a flash loan manipulation of price of one asset (like dai) to cause liquidation - value given by virtual price (if assumed to be in USD) is not affected by that manipulation

Reply:
virtual price by itself is "price of LP token assuming that the true price of any asset in the pool is 1.0"

Reply:
I didn't look deeply into the implementation, but it seems to me that the virtual price is something similar to the AAVE liquidity ratio, like `(deposits - withdrawals + fees) / (deposits - withdrawals)`, e.g. earnings ratio. Does it *really* incorporate current market prices information?
It seems that all the three points still stand.

Reply:
Ok, now I got the idea behind this approximation. It's a pity that a flash loan protected formal solution is not in sight. I'll take a bit more time to assess the approximation.

Reply:
It is definitely not like Aave's liquidity ratio actually.

The proof for an approximation is actually quite mathematically sound and easy

Reply:
https://drive.google.com/file/d/1PXk9uYZke-vQ4dCWu5gc35l9LFym6EK8/view?usp=sharing

Reply:
So if you have feeds for all stablecoins in LP, it is unnecessary to have just an oracle for lp

Reply:
Thanks for the proof!
We still need to ensure protection against flash loans in the case when assets behind the LP token are changed significantly and temporary. E.g., `sum(pool asset valuations) / Nt` won't work.

Is it possible to sway `Pv` significantly with a trade and then execute the reverse trade (assuming the fees and slippage economy makes the attack feasible)? As far as I can see there is no any TWAP-ness inside `Pv` that might help.